blog/2023-11-angular17/README.md

```diff
@@ -0,0 +1,427 @@
+
+Die Ausdrücke für den neuen Control Flow werden direkt im HTML-Code notiert und mit einem `@`-Symbol eingeleitet.
+
+
+### Bedingungen mit dem `@if`-Block
+
+Der `@if`-Block dient dazu, bestimmte Teile der Benutzeroberfläche nur dann anzuzeigen, wenn eine Bedingung erfüllt ist. Er ersetzt die Direktive `*ngIf`.
+Im `@if`-Block steht eine Bedingung. Nur wenn diese Bedingung wahr ist, wird der Teil der Benutzeroberfläche gezeigt.
+
+Ein `@if`-Block kann auch `@else`-Blöcke enthalten. Das sind alternative Blöcke, die angezeigt werden, wenn die Bedingung im `@if`-Teil nicht erfüllt ist.
+Man kann einen einfachen `@else`-Block haben, der immer dann zum Einsatz kommt, wenn die `@if`-Bedingung nicht zutrifft, oder man kann zusätzliche `@else`-Blöcke mit eigenen Bedingungen definieren:
+
+```html
+<!-- VORHER -->
+<ng-container *ngIf="books?.length >= 1; else elseBlock">
+  
+  <ng-container *ngIf="books?.length === 1; else elseBlock2">
+    <book-details [book]="books[0]" />
+  </ng-container>
+  
+  <ng-template #elseBlock2>
+    <book-list [books]="books" />
+  </ng-template>
+
+</ng-container>
+
+<ng-template #elseBlock>
+  <p>Keine Bücher verfügbar!</p>
+</ng-template>
+```
+
+```html
+<!-- NACHHER -->
+@if (books?.length > 1) {
+  <book-list [books]="books" />
+} @else if (books?.length === 1) {
+  <book-details [book]="books[0]" />
+} @else {
+  <p>Keine Bücher verfügbar!</p>
+}
+```
+
+Es fällt sofort auf, dass der `else`-Zweig deutlich einfacher zu definieren ist als zuvor. Der Einsatz von `<ng-template>` ist nicht mehr notwendig. Bislang war es häufig erforderlich, das spezielle Element `<ng-container>` einzusetzen, um mehrere Elemente zu gruppieren, ohne ein unnötiges DOM-Element zu erzeugen.
+Auch dies entfällt mit der neuen Syntax, da die Gruppierung nun über die Klammern geschieht.
+
+
+### Wiederholungen mit `@for`
+
+Der Schleifenblock ersetzt `*ngFor` für Iterationen und unterscheidet sich in einigen Punkten von der bislang eingesetzten Direktive `*ngFor`-Direktive. Der neue Schleifenblock sieht so aus:
+
+```html
+<!-- VORHER --> 
+<ul class="names">
+  <li *ngFor="book of books">
+    <book-list-item [book]="book" />
+  </li>
+  <p *ngIf="!books.length">Keine Bücher verfügbar!</p>
+</ul>
+```
+
+```html
+<!-- NACHHER --> 
+<ul class="names">
+  @for (book of books; track book.isbn) {
+    <li>
+      <book-list-item [book]="book" />
+    </li>
+  } @empty {
+    <li>Keine Bücher verfügbar!</li>
+  }
+</ul>
+```
+
+Es ist nun möglich, direkt einen `@empty`-Block anzugeben, der aktiv wird, wenn es keine Einträge gibt. Dies war zuvor nicht direkt möglich.
+Die gezeigte Einstellung `track` ersetzt das Konzept der `trackBy`-Funktion. Sie bestimmt den Schlüssel für jede Zeile, den der Schleifenblock benutzt, um Array-Elemente eindeutig zu identifizieren. Zuvor war es optional, das explizite Tracking zu verwenden, mit der neuen Syntax ist `track` jedoch eine Pflichtangabe.
```

Comment:
Vielleicht hier noch den Hinweis, dass zuvor implizit der Index genutzt wurde und man das jetzt auch über $index wieder machen könnte / kann, wenn kein sinnvolles track-attribut vorliegt?

Reply:
In dem Fall soll man den Wert verwenden. Ich hab dafür weiter unten ein Beispiel hinzugefügt: `@for (name of nameList; track name)`

Reply:
Ist das ausreichend oder willst du noch mehr Infos dazu?

Reply:
Ah okay, nein das reicht mir.